### PR TITLE
Reload BasicUI if sitemap has been changed

### PR DESCRIPTION
**Pull request by triller-telekom**
This catches the IndexOutofBoundException from #3836

Signed-off-by: Stefan Triller <stefan.triller@telekom.de>

### CODE REVIEW

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js [hunk 1]
```diff
@@ -1709,6 +1709,16 @@
 				value,
 				title;
 
+			if(data.TYPE === "SITEMAP_CHANGED") {
```
Comment (sjka) [inline extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: isn't it also necessary to check if it is _this_ sitemap which changed? I.e. have you tried having to browser windows open with different sitemaps and check that only the affected one gets redirected to its main page?

File: bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/PageChangeListener.java [hunk 2]
```diff
@@ -213,8 +213,13 @@ private boolean definesVisibility(Widget w, String name) {
         return false;
     }
 
-    public void sitemapContentChanged(EList<Widget> widgets) {
-        updateItemsAndWidgets(widgets);
+    public void sitemapContentChanged() {
+        for (SitemapSubscriptionCallback callback : distinctCallbacks) {
+            SitemapChangedEvent changeEvent = new SitemapChangedEvent();
```
Comment (sjka) [inline bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/PageChangeListener.java]: Does it really make sense to create a new event for each callback and not create it once and send it to all of them?

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js (hunk 1 above)
Comment (triller-telekom) [inline extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: Nothing :) Grunt expects this format or it complains and does not generate the minified version of the file

Comment (triller-telekom) [inline extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: While fixing this bug I noticed that its almost impossible to detect in which page the user currently is if the the order of the pages has been changed. Thats why there is a reload for every sitemap change, also for browsers which are on the first page. Just replacing the widgets would have let to inconsistent views in the browser. This solution is now much cleaner.

File: bundles/ui/org.eclipse.smarthome.ui/src/main/java/org/eclipse/smarthome/ui/internal/items/ItemUIRegistryImpl.java [hunk 3]
```diff
@@ -515,7 +515,7 @@ public Widget getWidget(Sitemap sitemap, String id) {
                     for (int i = 2; i < id.length(); i += 2) {
                         w = ((LinkableWidget) w).getChildren().get(Integer.valueOf(id.substring(i, i + 2)));
                     }
-                } catch (NumberFormatException e) {
+                } catch (NumberFormatException | IndexOutOfBoundsException e) {
```
Comment (triller-telekom) [inline bundles/ui/org.eclipse.smarthome.ui/src/main/java/org/eclipse/smarthome/ui/internal/items/ItemUIRegistryImpl.java]: Alright, I will change it. Though the `NumberFormatException` was caught here I thought it doesn't hurt to catch the other one too...

File: bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/PageChangeListener.java (hunk 2 above)
Comment (triller-telekom) [inline bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/PageChangeListener.java]: Agreed, creating one event is sufficient.

**Comment by sjka** [comment]
okay from my side.
@resetnow your call.

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js [hunk 4]
```diff
@@ -1709,6 +1709,16 @@
...
 				title;
 
+			if (data.TYPE === "SITEMAP_CHANGED") {
+				var oldLocation = window.location.href;
+				var parts = oldLocation.split("?");
+				if (parts.length > 1) {
+					window.location.href = parts[0] + "?sitemap="+data.sitemapName;
+				} else {
+					window.location.href = oldLocation;
+				}
```
Comment (resetnow) [inline extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]:
```

_t.pause();
return;
```

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js [hunk 5]
```diff
@@ -1709,6 +1709,16 @@
 				value,
 				title;
 
+			if (data.TYPE === "SITEMAP_CHANGED") {
+				var oldLocation = window.location.href;
+				var parts = oldLocation.split("?");
+				if (parts.length > 1) {
+					window.location.href = parts[0] + "?sitemap="+data.sitemapName;
```
Comment (resetnow) [inline extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: Spaces:

```
"?sitemap=" + data
```

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js [hunk 6]
```diff
@@ -1709,6 +1709,16 @@
...
+				} else {
+					window.location.href = oldLocation;
```
Comment (resetnow) [inline extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]:
```
window.location.reload(true);
```

**Comment by kaikreuzer** [comment]
> This catches the IndexOutofBoundException from #3836 

Does it mean that #3836 is hence fixed now and can be closed?

**Comment by triller-telekom** [comment]
Yes it can be closes because the sitemap problem is solved and the other problem regarding the non working D-Link camera has been extracted into #3870

**Pull request merged**